src/main/java/dk/cngroup/lentils/service/StatisticsService.java

```diff
@@ -0,0 +1,85 @@
+package dk.cngroup.lentils.service;
+
+import dk.cngroup.lentils.entity.Cypher;
+import dk.cngroup.lentils.entity.CypherStatus;
+import dk.cngroup.lentils.entity.HintTaken;
+import dk.cngroup.lentils.repository.CypherRepository;
+import dk.cngroup.lentils.repository.HintTakenRepository;
+import dk.cngroup.lentils.repository.StatusRepository;
+import dk.cngroup.lentils.repository.StatusRepositorySpec;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Service;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.LinkedHashMap;
+
+@Service
+public class StatisticsService {
+    @Autowired
+    private StatusRepository statusRepository;
+
+    @Autowired
+    private CypherRepository cypherRepository;
+
+    @Autowired
+    private HintTakenRepository hintTakenRepository;
+
+    public Map<Cypher, Integer> getCountOfSpecificStatusesOfCyphers(final CypherStatus cypherStatus) {
+
+        Map<Cypher, Integer> countOfStatusesOfCyphers = new HashMap<>();
+        List<Cypher> allCyphers = cypherRepository.findAll();
+
+        for (Cypher cypher : allCyphers) {
+            countOfStatusesOfCyphers
+                    .put(cypher, statusRepository.findAll(Specification
+                                    .where(StatusRepositorySpec.hasCypher(cypher)
+                                    .and(StatusRepositorySpec.hasCypherStatus(cypherStatus))))
+                            .size());
+        }
+        return sortMapByValueDesc(countOfStatusesOfCyphers);
+    }
+
+    public Map<Cypher, Integer> getNumberOfHintsTakenOfCyphers() {
+
+        Map<Cypher, Integer> hintsTakenOfCyphers = new HashMap<>();
+        List<Cypher> cyphers = cypherRepository.findAll();
+
+        for (Cypher cypher : cyphers) {
+            hintsTakenOfCyphers.put(cypher, hintTakenRepository.findAllByCypher(cypher.getCypherId()).size());
+        }
+        return sortMapByValueDesc(hintsTakenOfCyphers);
+    }
+
+    public Map<Cypher, Integer> getPointsOfHintsTakenOfCyphers() {
+
+        Map<Cypher, Integer> hintsTakenOfCyphers = new HashMap<>();
+        List<Cypher> cyphers = cypherRepository.findAll();
+
+        for (Cypher cypher : cyphers) {
+            List<HintTaken> hintsTaken = hintTakenRepository.findAllByCypher(cypher.getCypherId());
+
+            int hintsValue = 0;
```

Comment:
Tady to uz elegantnejsi je, takze bych prepsal na:
```java
int hintsValue = hintsTaken.stream().mapToInt(ht -> ht.getHint().getValue()).sum();
```
Mimochodem tohle by mohla byt somostatna metoda v nejake slozbe.

Reply:
extrahoval jsem